openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -469,45 +473,70 @@ def _verify_user(request, response, userid_in_session):
 
             if request_user_object_mismatch or session_user_mismatch:
                 # Log accumulated information stored on request for each change of user
-                if hasattr(request, 'debug_user_changes'):
-                    log.warning('An unsafe user transition was found. It either needs to be fixed or exempted.\n {}'
-                                .format('\n'.join(request.debug_user_changes)))
+                extra_logs = []
+
+                # One of the possibilities for why we get an occasional
+                # safe-session user mismatch is that the wrong session is
+                # retrieved from cache. This additional logging should reveal
+                # any such mismatch (without revealing the actual session ID
+                # in logs).
+                session_hashes = [
+                    ('cookie_field', obscure_token(request.session_cookie_string[2:34])),
+                    ('at_request', obscure_token(request.safe_cookie_verified_session_id)),
+                    ('at_response', obscure_token(request.session.session_key)),
+                ]
 
                 session_id_changed = hasattr(request.session, 'session_key') and\
```

Comment:
Would it be more clear if this ensured that all session_hashes were equivalent? And if not hashes, should this be checking the same 3 actual values, rather than just 2 of them?

Reply:
Mmm, I could update it -- or just remove it, if we're going to print the hashes anyhow. (Or either print "they're all the same" if they match, and the values if they don't...)

But I also see that I've written `request.session.session_key` and `request.session` might be `None` in some cases, so I should fix that too.

Reply:
My preference would be to update the language, but to check the hashes for matches and call that out in the message, because it provides something to search for.

Reply:
Could we also add something like?:
```
set_custom_attribute(f'safe_sessions.session_id_changed', session_id_changed)
```

Reply:
True, something to search for is good.